lib/internal/modules/esm/loader.js

```diff
@@ -594,56 +719,100 @@ class ESMLoader {
       ['string', 'URL'],
       parentURL,
     );
+    const resolvers = this.#resolvers;
+
+    let i = resolvers.length - 1;
+    let {
+      url: resolverFilePath,
+      fn: resolver,
+    } = resolvers[i];
+    let chainFinished = i === 0;
```

Comment:
Same notes as above per `load`. Could we abstract this “work through all the user-loaded hooks” logic and share it between `resolve` and `load`? (I’m assuming `globalPreload` is probably too different, since it concatenates its output rather than having one hook replace another hook’s output, but if it’s not too confusing to include that one too we could.)

Reply:
I already went down that road. I think unfortunately no (or at least not without creating more complexity than it solves): these _look_ very similar (and certainly use the same mechanic), but there are too many inputs, hoisted references, and the guts of each `next()` are actually very different. It ends up a Frankenstein. I can take another look at it if you'd really like; I abandoned it pretty quickly the first time because it quickly spiralled.

Yes, it does not apply to `globalPreload` for exactly that reason 🙂

Reply:
Ah, I forgot that the return values also differ. Good point.

I guess there’s no way to abstract just the “iterate through the hooks” part, since that would seem to be identical for all of them? No, because that can’t be easily separated from dealing with return values?

Reply:
I think the biggest challenge is the guts of `next()` which are very different aside from the counting down. It depends on reading and overwriting variables from a parent scope, and abstracting that away is medicine worse than the disease :(

Reply:
Fair enough. I’m tempted to say that this almost sounds like one of the few use cases for writing a generator function but that also sounds worse.